### ISSUE TITLE
Generic interfaces for MatchResults, route state, query, etc

### ISSUE THREAD

**Issue by AkhmedovValikhan** [ionic-team/stencil-router#89]
**Resources:**
Before submitting an issue, please consult our [docs](https://stenciljs.com/).

**Stencil version:** (run `npm list @stencil/core` from a terminal/cmd prompt and paste output below):

```
 insert the output from npm list @stencil/core here
```

**I'm submitting a ...**  (check one with "x")
[ ] bug report
[x] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or https://stencil-worldwide.slack.com

**Current behavior:**
It would be really nice to have generic versions of router interfaces such as MathResults, location query, state, hash. So you can strongly type those objects, currenly I am doing something like this, it works fine, but looks a bit unfriendly
![image](https://user-images.githubusercontent.com/9393189/52918392-e7eb6300-330f-11e9-9551-186c6b1c2725.png)
So basically I can now use it in my component wich has @Prop() match: MyTimeoffMatch, and I get intelliscene and type checking
**Expected behavior:**

**Steps to reproduce:**

**Related code:**

```
insert any relevant code here
```

**Other information:**
